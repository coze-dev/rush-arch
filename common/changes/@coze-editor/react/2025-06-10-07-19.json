{
  "changes": [
    {
      "packageName": "@coze-editor/react",
      "comment": "tmp",
      "type": "patch"
    }
  ],
  "packageName": "@coze-editor/react",
  "email": "tecvan.fe@qq.com"
}
