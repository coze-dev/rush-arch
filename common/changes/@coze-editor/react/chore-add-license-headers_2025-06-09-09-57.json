{
  "changes": [
    {
      "packageName": "@coze-editor/react",
      "comment": "add license header by running rush license-header",
      "type": "patch"
    }
  ],
  "packageName": "@coze-editor/react",
  "email": "stream-pipe@users.noreply.github.com"
}
