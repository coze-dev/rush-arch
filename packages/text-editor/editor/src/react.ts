//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

export * from '@coze-editor/react';
export * from '@coze-editor/react-components';
export * from '@coze-editor/react-hooks';
