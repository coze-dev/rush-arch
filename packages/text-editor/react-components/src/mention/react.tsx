//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

import { useInjectorEffect, useLatest } from '@coze-editor/react-hooks';

import { type MentionOptions } from './types';
import { hasTrigger } from './is';
import { mention, getCurrentMentionReplaceRange } from './extension';

function Mention(props: MentionOptions) {
  const propsRef = useLatest(props);

  useInjectorEffect(injector => {
    const sharedOptions: Partial<MentionOptions> = {
      search: props.search,
      onOpenChange(...args) {
        if (typeof propsRef.current.onOpenChange === 'function') {
          return propsRef.current.onOpenChange(...args);
        }
      },
      onSearch(...args) {
        if (typeof propsRef.current.onSearch === 'function') {
          return propsRef.current.onSearch(...args);
        }
      },
    };
    return injector.inject([
      mention(
        hasTrigger(props)
          ? {
              ...sharedOptions,
              trigger(tr) {
                if (hasTrigger(propsRef.current)) {
                  return propsRef.current.trigger(tr);
                }
              },
            }
          : {
              ...sharedOptions,
              triggerCharacters: props.triggerCharacters ?? [],
            },
      ),
    ]);
  });

  return null;
}

export { Mention, getCurrentMentionReplaceRange };

export type { MentionOptions };
