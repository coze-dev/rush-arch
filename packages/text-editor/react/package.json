{
  "name": "@coze-editor/react",
  "version": "1.0.0-alpha.45",
  "description": "react",
  "license": "MIT",
  "author": "fengzilong",
  "maintainers": [],
  "sideEffects": [
    "**/*.css",
    "**/*.less",
    "**/*.sass",
    "**/*.scss"
  ],
  "main": "./dist/esm/index.js",
  "module": "./dist/esm/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "dev": "vite --force",
    "lint": "eslint && tsc --noEmit",
    "prepublishOnly": "pnpm run build"
  },
  "dependencies": {
    "@coze-editor/core": "workspace:*"
  },
  "devDependencies": {
    "@codemirror/state": "^6.4.1",
    "@codemirror/view": "^6.26.1",
    "@coze-arch/ts-config": "workspace:*",
    "@coze-editor/eslint-config": "workspace:*",
    "@types/node": "^22",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "eslint": "9.14.0",
    "react": "~18.2.0",
    "react-dom": "~18.2.0",
    "tsup": "^8.0.1",
    "typescript": "^5.8.2"
  },
  "peerDependencies": {
    "@codemirror/state": "^6.4.1",
    "@codemirror/view": "^6.26.1",
    "react": "~18.2.0",
    "react-dom": "~18.2.0"
  },
  "packageManager": "pnpm@9.15.0",
  "test:main": "./src/index.ts"
}
