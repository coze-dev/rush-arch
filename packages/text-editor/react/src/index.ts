export { Renderer, type InferRendererProps } from './renderer';

export { createRenderer } from './create-renderer';

export { EditorProvider, useEditor, useInjector } from './provider';

export * from '@coze-editor/core';
