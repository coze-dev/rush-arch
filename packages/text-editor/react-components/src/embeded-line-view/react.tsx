//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

import { createPortal } from 'react-dom';
import { forwardRef, type ReactNode, useImperativeHandle } from 'react';

import {
  useHTMLElement,
  useInjectorEffect,
  useLatest,
} from '@coze-editor/react-hooks';
import { type BuiltinEditorAPI, useEditor } from '@coze-editor/react';
import { Decoration, EditorView, WidgetType } from '@codemirror/view';
import { MapMode, StateEffect, StateField } from '@codemirror/state';

interface EmbededLineViewProps {
  children?: ReactNode;
}

class DOMWidget extends WidgetType {
  constructor(private options: { dom: HTMLElement }) {
    super();
  }

  toDOM() {
    return this.options.dom;
  }

  eq(other: DOMWidget) {
    return this.options.dom === other.options.dom;
  }
}

interface UpdateOptions {
  visible: boolean;
  // line number, 1-based, defaults to 1
  line?: number;
  // where to put the view, before or after the target line
  side?: EmbededLineViewSide;
}

interface EmbededLineViewRefProps {
  update: (options: UpdateOptions) => void;
}

enum EmbededLineViewSide {
  Before = 'before',
  After = 'after',
}

const updateEffect = StateEffect.define<Required<UpdateOptions>>();

const EmbededLineView = forwardRef<
  EmbededLineViewRefProps,
  EmbededLineViewProps
>(function EmbededLineView({ children }, ref) {
  const dom = useHTMLElement('div');
  const domRef = useLatest(dom);
  const editor = useEditor<BuiltinEditorAPI>();
  const editorRef = useLatest(editor);

  useImperativeHandle(ref, () => ({
    update({ visible, line, side }: UpdateOptions) {
      editorRef.current?.$view.dispatch({
        effects: updateEffect.of({
          visible,
          line: line ?? 1,
          side: side ?? EmbededLineViewSide.Before,
        }),
      });
    },
  }));

  useInjectorEffect(injector => {
    const field = StateField.define({
      create() {
        return {
          position: 0,
          side: EmbededLineViewSide.Before,
          decorations: Decoration.none,
        };
      },
      update(value, tr) {
        const newPosition = tr.changes.mapPos(
          value.position,
          -1,
          MapMode.Simple,
        );

        let newValue = value;
        if (typeof newPosition === 'number' && newPosition !== value.position) {
          const docLine = tr.state.doc.lineAt(newPosition);

          let decoPos = 0;

          if (value.side === EmbededLineViewSide.After) {
            decoPos = docLine.to;
          } else {
            decoPos = docLine.from;
          }

          newValue = {
            ...value,
            position: newPosition,
            decorations: Decoration.set([
              Decoration.widget({
                widget: new DOMWidget({
                  dom: domRef.current,
                }),
                block: true,
                side: value.side === EmbededLineViewSide.After ? 1 : -1,
              }).range(decoPos),
            ]),
          };
        }

        for (const effect of tr.effects) {
          if (effect.is(updateEffect)) {
            const { visible, line, side } = effect.value;

            if (!visible) {
              return {
                position: newValue.position,
                side,
                decorations: Decoration.none,
              };
            }

            // out of range, ignore
            if (line < 1 || line > tr.startState.doc.lines) {
              continue;
            }

            const docLine = tr.startState.doc.line(line);

            let decoPos = 0;

            if (side === EmbededLineViewSide.After) {
              decoPos = docLine.to;
            } else {
              decoPos = docLine.from;
            }

            return {
              position: decoPos,
              side,
              decorations: Decoration.set([
                Decoration.widget({
                  widget: new DOMWidget({
                    dom: domRef.current,
                  }),
                  block: true,
                  side: newValue.side === EmbededLineViewSide.After ? 1 : -1,
                }).range(decoPos),
              ]),
            };
          }
        }

        return newValue;
      },
      provide(f) {
        return EditorView.decorations.compute(
          [f],
          state => state.field(f).decorations,
        );
      },
    });

    return injector.inject([field]);
  }, []);

  return createPortal(children, dom);
});

export { EmbededLineView, EmbededLineViewSide, type EmbededLineViewRefProps };
