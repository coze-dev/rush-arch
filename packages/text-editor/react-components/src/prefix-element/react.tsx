import { createPortal } from 'react-dom';
import { type ReactNode, useLayoutEffect } from 'react';

import { useHTMLElement, useLatest } from '@coze-editor/react-hooks';
import { useInjector } from '@coze-editor/react';
import { Decoration, EditorView, keymap, WidgetType } from '@codemirror/view';

import { clientRectsFor, flattenRect } from './dom';

class PrefixElementWidget extends WidgetType {
  constructor(
    readonly content:
      | string
      | HTMLElement
      | ((view: EditorView) => HTMLElement),
  ) {
    super();
  }

  toDOM(view: EditorView) {
    const wrap = document.createElement('span');
    wrap.className = 'cm-prefix-element';
    // solved the cursor height problem(eg. cursor double height with two lines)
    wrap.style.cssText = 'height: 0;';
    wrap.appendChild(
      typeof this.content === 'string'
        ? document.createTextNode(this.content)
        : typeof this.content === 'function'
          ? this.content(view)
          : this.content,
    );
    if (typeof this.content === 'string') {
      wrap.setAttribute('aria-label', `prefix ${this.content}`);
    } else {
      wrap.setAttribute('aria-hidden', 'true');
    }
    return wrap;
  }

  coordsAt(dom: HTMLElement) {
    const rects = dom.firstChild ? clientRectsFor(dom.firstChild) : [];
    if (!rects.length) {
      return null;
    }
    const style = window.getComputedStyle(dom.parentNode as HTMLElement);
    const rect = flattenRect(rects[0], style.direction != 'rtl');
    const lineHeight = parseInt(style.lineHeight);
    if (rect.bottom - rect.top > lineHeight * 1.5) {
      return {
        left: rect.left,
        right: rect.right,
        top: rect.top,
        bottom: rect.top + lineHeight,
      };
    }
    return rect;
  }

  ignoreEvent() {
    return false;
  }
}

interface PrefixElementProps {
  deletable?: boolean;
  onDelete?: () => void;
  children?: ReactNode;
}

function PrefixElement({ deletable, onDelete, children }: PrefixElementProps) {
  const element = useHTMLElement('span');
  const latestElement = useLatest(element);
  const latestDeletable = useLatest(deletable);
  const latestOnDelete = useLatest(onDelete);
  const injector = useInjector();

  useLayoutEffect(() => {
    function run(view: EditorView) {
      if (
        view.state.selection.main.empty &&
        view.state.selection.main.head === 0 &&
        latestDeletable.current === true &&
        typeof latestOnDelete.current === 'function'
      ) {
        latestOnDelete.current();
      }

      return false;
    }

    return injector.inject([
      EditorView.decorations.of(
        Decoration.set([
          Decoration.widget({
            side: -100,
            block: false,
            widget: new PrefixElementWidget(() => latestElement.current),
          }).range(0),
        ]),
      ),
      keymap.of([
        { key: 'Backspace', shift: run, run },
        { key: 'Meta-Backspace', shift: run, run },
      ]),
    ]);
  }, []);

  return createPortal(children, element);
}

export { PrefixElement };
