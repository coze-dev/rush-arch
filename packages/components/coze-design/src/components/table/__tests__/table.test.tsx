//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

import { describe, it } from 'vitest';
import { render, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';

import { LocaleProviderContainer } from '@/__tests__/locale-provider-containner';

import { Table } from '..';

const { TableMeta, TableAction } = Table;

describe('Table', () => {
  it('should has className coz-table-wrapper', () => {
    const { container } = render(<Table />);
    expect(container.getElementsByClassName('coz-table-wrapper')).toHaveLength(
      1,
    );
  });

  it('should has status empty', () => {
    const { container } = render(<Table tableProps={{ dataSource: [] }} />);
    expect(container.getElementsByClassName('coz-empty-content')).toHaveLength(
      1,
    );
  });

  it('should has status loading', () => {
    const { container } = render(<Table tableProps={{ loading: true }} />);
    expect(container.getElementsByClassName('coz-table-spin')).toHaveLength(1);
  });

  it('table data ready', () => {
    const { container } = render(
      <LocaleProviderContainer>
        <Table
          tableProps={{ dataSource: [{ name: '123' }] }}
          enableLoad
          loadMode="cursor"
          hasMore={true}
        />
      </LocaleProviderContainer>,
    );
    expect(container.getElementsByClassName('coz-table-list')).toHaveLength(1);
  });

  it('should has className coz-table-meta', () => {
    const { container } = render(
      <TableMeta
        icon_url="test table meta icon url"
        name="test table meta name"
        description="test table meta description"
      />,
    );
    expect(container.getElementsByClassName('coz-table-meta')).toHaveLength(1);
  });

  it('should has className table meta no avatar', () => {
    const { container } = render(
      <TableMeta
        icon_url=""
        name="table meta withtou avatar"
        description="test table meta description"
      />,
    );
    expect(container.getElementsByClassName('semi-avatar')).toHaveLength(0);
  });

  it('should has className table action', () => {
    const { container } = render(
      <LocaleProviderContainer>
        <TableAction
          deleteProps={{ disabled: false }}
          copyProps={{ disabled: false }}
          editProps={{ disabled: false }}
          actionList={[
            {
              customRender: <div className="test-custom-render"></div>,
            },
            {
              actionKey: 'enable',
              actionText: 'enable',
              disabled: false,
            },
          ]}
        />
      </LocaleProviderContainer>,
    );
    expect(container.getElementsByClassName('coz-table-action')).toHaveLength(
      1,
    );
  });

  describe('indexRowSelection', () => {
    it('should has className coz-table-list-index-row-selection', () => {
      const { container } = render(
        <LocaleProviderContainer>
          <Table
            indexRowSelection
            tableProps={{ dataSource: [{ name: '123', key: 123 }] }}
          />
        </LocaleProviderContainer>,
      );
      expect(
        container.getElementsByClassName('coz-table-list-index-row-selection'),
      ).toHaveLength(1);
    });

    it('should render checkbox cell with index and checkbox', () => {
      const { container } = render(
        <LocaleProviderContainer>
          <Table
            indexRowSelection
            tableProps={{ dataSource: [{ name: '123', key: 123 }] }}
          />
        </LocaleProviderContainer>,
      );
      expect(
        container.getElementsByClassName('coz-table-checkbox-cell'),
      ).toHaveLength(2);

      const checkboxCell = container.getElementsByClassName(
        'coz-table-checkbox-cell',
      );

      Array.from(checkboxCell).forEach(cell => {
        expect(cell.innerHTML).toMatchSnapshot();
      });
    });

    it('should handle click event', () => {
      const { container } = render(
        <LocaleProviderContainer>
          <Table
            indexRowSelection
            tableProps={{
              dataSource: [
                { name: '123', key: '123' },
                { name: '456', key: '456' },
              ],
              columns: [{ dataIndex: 'name', key: 'name' }],
            }}
          />
        </LocaleProviderContainer>,
      );

      const checkboxInput = container.querySelector(
        '.coz-table-checkbox-cell-checkbox input[type="checkbox"]',
      );

      if (!checkboxInput) {
        throw new Error('Checkbox input not found');
      }

      act(() => {
        fireEvent.click(checkboxInput);
      });

      // 检查 checkbox 是否变为选中状态
      const checkbox = container.querySelector(
        '.coz-table-checkbox-cell-checkbox .semi-checkbox',
      );
      expect(checkbox).toHaveClass('semi-checkbox-checked');

      // 检查列选择框是否变为中间状态
      const columnSelection = container.querySelector(
        '.semi-table-column-selection .semi-checkbox',
      );
      console.log('columnSelection', columnSelection);
      expect(columnSelection).toHaveClass('semi-checkbox-indeterminate');
    });
  });
});
