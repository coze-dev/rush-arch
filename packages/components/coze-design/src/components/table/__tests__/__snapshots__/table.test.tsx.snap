// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > indexRowSelection > should render checkbox cell with index and checkbox 1`] = `"<span class="semi-checkbox semi-checkbox-unChecked semi-checkbox-cardType_enable coz-checkbox"><span class="semi-checkbox-inner"><input type="checkbox" aria-label="Select all rows" aria-disabled="false" aria-checked="false" class="semi-checkbox-input"><span class="semi-checkbox-inner-display"></span></span></span>"`;

exports[`Table > indexRowSelection > should render checkbox cell with index and checkbox 2`] = `"<span class="coz-table-checkbox-cell-index coz-fg-secondary text-sm font-normal">1</span><span class="coz-table-checkbox-cell-checkbox"><span class="semi-checkbox semi-checkbox-unChecked semi-checkbox-cardType_enable coz-checkbox"><span class="semi-checkbox-inner"><input type="checkbox" aria-label="Select this row" aria-disabled="false" aria-checked="false" class="semi-checkbox-input"><span class="semi-checkbox-inner-display"></span></span></span></span>"`;
